src/compile/scale/assemble.ts

```diff
@@ -5,47 +5,51 @@ import {stringValue, vals} from '../../util';
 import {Model} from '../model';
 import {isRawSelectionDomain, selectionScaleDomain} from '../selection/selection';
+import {mergeDomains} from './domain';
 
 
 export function assembleScale(model: Model): VgScale[] {
-    return vals(model.component.scales).reduce((scales: VgScale[], scaleComponent) => {
+    return vals(model.component.scales).reduce((scales, scaleComponent) => {
       if (scaleComponent.merged) {
         // Skipped merged scales
         return scales;
       }
 
-      // We need to cast here as combine returns Partial<VgScale> by default.
-      const scale = scaleComponent.combine(['name', 'type', 'domain', 'domainRaw', 'range']) as VgScale;
+      const {name, type, domainRaw: raw, range, ...otherScaleProps} = scaleComponent.combine();
+      let domainRaw = raw;  // needs to be mutable
 
-      const domainRaw = scaleComponent.get('domainRaw');
       // As scale parsing occurs before selection parsing, a temporary signal
       // is used for domainRaw. Here, we detect if this temporary signal
       // is set, and replace it with the correct domainRaw signal.
       // For more information, see isRawSelectionDomain in selection.ts.
       if (domainRaw && isRawSelectionDomain(domainRaw)) {
-        scale.domainRaw = selectionScaleDomain(model, domainRaw);
+        domainRaw = selectionScaleDomain(model, domainRaw);
       }
 
-      // Correct references to data as the original domain's data was determined
-      // in parseScale, which happens before parseData. Thus the original data
-      // reference can be incorrect.
-      const domain = scaleComponent.get('domain');
-      if (isDataRefDomain(domain) || isFieldRefUnionDomain(domain)) {
-        domain.data = model.lookupDataSource(domain.data);
-        scales.push(scale);
-      } else if (isDataRefUnionedDomain(domain)) {
-        domain.fields = domain.fields.map((f: VgDataRef) => {
-          return {
-            ...f,
-            data: model.lookupDataSource(f.data)
-          };
-        });
-        scales.push(scale);
-      } else if (isSignalRefDomain(domain) || isArray(domain)) {
-        scales.push(scale);
-      } else {
-        throw new Error('invalid scale domain');
-      }
+      const domains = scaleComponent.domains.map(domain => {
+        // Correct references to data as the original domain's data was determined
+        // in parseScale, which happens before parseData. Thus the original data
+        // reference can be incorrect.
+
+        if (isDataRefDomain(domain)) {
+          domain.data = model.lookupDataSource(domain.data);
+        }
+        return domain;
+      });
+
+      // domains is an array that has to be merged into a single vega domain
+      const domain = mergeDomains(domains);
+
+
+      scales.push({
+        name,
+        type,
+        domain: domain,
+        range: range,
+        ...(domainRaw ? {domainRaw} : {}),
```

Comment:
Actually, domainRaw should come before range.  (My previous comment might be wrong too.)